File: python/taflow/metrics/maximum_drawdown_duration.py
"""Longest observation-count drawdown episode metric."""
from __future__ import annotations
from typing import Any
from .._native.metrics import MaximumDrawdownDuration as _Native
from ._input import as_metric_series


class MaximumDrawdownDuration:
    """Return the longest drawdown length in observations.

    The oracle contract is PerformanceAnalytics 2.1.0 ``findDrawdowns`` with
    geometric wealth: only negative-drawdown episodes qualify, and length
    includes the peak/recovery boundary, so one underwater observation has
    length two. An unrecovered current episode uses the same convention.
    Warm-up ends at the first negative drawdown; a path that never falls below
    its running peak returns ``None``. NaNs are omitted by default. Rust owns
    the O(1) state and converts returns, log returns, equity, or period P&L.
    """
    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(self, returns: Any, *, column: str | None = None) -> "MaximumDrawdownDuration":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_log_returns(self, log_returns: Any, *, column: str | None = None) -> "MaximumDrawdownDuration":
        """Append chronological log returns observations and return this metric."""
        self._state.from_log_returns(as_metric_series(log_returns, column=column))
        return self

    def from_equity(self, equity: Any, *, column: str | None = None) -> "MaximumDrawdownDuration":
        """Append chronological equity observations and return this metric."""
        self._state.from_equity(as_metric_series(equity, column=column))
        return self

    def from_pnl(self, pnl: Any, initial_capital: float, *, column: str | None = None) -> "MaximumDrawdownDuration":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column), float(initial_capital))
        return self

    def append(self, value: float) -> "MaximumDrawdownDuration":
        """Append one observation and return this metric."""
        self._state.append(float(value)); return self
    def extend(self, values: Any, *, column: str | None = None) -> "MaximumDrawdownDuration":
        """Append chronological observations and return this metric."""
        self._state.extend(as_metric_series(values, column=column)); return self
    @property
    def value(self) -> int | None:
        """Return the maximum episode length, or ``None`` when undefined."""
        return self._state.value
    def compute(self) -> int | None:
        """Return the current scalar without replaying input."""
        return self._state.compute()
    def reset(self) -> "MaximumDrawdownDuration":
        """Clear observations, preserve settings, and return this metric."""
        self._state.reset(); return self
    def __len__(self) -> int:
        """Return usable normalized-return count delegated to Rust."""
        return len(self._state)


__all__ = ["MaximumDrawdownDuration"]

File: crates/taflow-metrics/src/metrics/maximum_drawdown_duration.rs
use crate::{
    primitives::DrawdownState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Longest drawdown episode length in usable observations.
#[derive(Debug, Clone)]
pub struct MaximumDrawdownDuration {
    input: MetricInputState,
    drawdown: DrawdownState,
    underwater_observations: usize,
    maximum_duration: usize,
}

impl MaximumDrawdownDuration {
    /// Construct an empty path-duration state.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            drawdown: DrawdownState::new(),
            underwater_observations: 0,
            maximum_duration: 0,
        })
    }

    /// Append one chronological observation and return the longest episode length.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<usize>> {
        if let Some(simple_return) = self.input.append(value)? {
            self.drawdown.append(simple_return)?;
            let underwater = self
                .drawdown
                .current_drawdown()
                .is_some_and(|drawdown| drawdown < 0.0);
            if underwater {
                self.underwater_observations += 1;
                self.maximum_duration = self.maximum_duration.max(self.underwater_observations + 1);
            } else if self.underwater_observations != 0 {
                self.maximum_duration = self.maximum_duration.max(self.underwater_observations + 1);
                self.underwater_observations = 0;
            }
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<usize>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }

    /// Return the longest negative drawdown run including its peak boundary.
    pub fn value(&self) -> Option<usize> {
        (self.maximum_duration != 0).then_some(self.maximum_duration)
    }
    /// Return the current result without replaying input.
    pub fn compute(&self) -> Option<usize> {
        self.value()
    }
    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.drawdown.reset();
        self.underwater_observations = 0;
        self.maximum_duration = 0;
    }
    /// Return the number of usable normalized returns processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }
    /// Return whether no usable returns have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_return_metric_lifecycle!(MaximumDrawdownDuration);
